core/src/main/java/net/opentsdb/query/processor/merge/Merger.java

```diff
@@ -81,54 +149,215 @@ public QueryNodeConfig config() {
 
   @Override
   public void close() {
-    // No-op
+    for (final Entry<QueryResultId, Waiter> entry : results.entrySet()) {
+      final Waiter waiter = entry.getValue();
+      if (waiter.timeout != null) {
+        waiter.timeout.cancel();
+      }
+    }
   }
   
   @Override
   public void onNext(final QueryResult next) {
-    final String id = next.source().config().getId();
+    if (complete.get()) {
+      // lost a race so don't bother.
+      return;
+    }
+
+    int countdown = 0;
+    Waiter extant = results.get(next.dataSource());
+    if (extant == null) {
+      LOG.warn("Received a result in node " + config.getId()
+              + " for an unexpected result " + next.dataSource()
+              + ". Expecting " + config.sortedSources());
+      return;
+    }
+
+    if (extant.result != null) {
+      sendUpstream(new IllegalStateException("Received a duplicate result for "
+              + next.dataSource() + " in node " + config.getId()));
+      return;
+    }
+
+    extant.result = next;
+
+    if (extant.timeout != null) {
+      extant.timeout.cancel();
+      extant.timeout = null;
+    }
+    countdown = outstanding.decrementAndGet();
+
+    if (countdown != 0) {
+      if (LOG.isTraceEnabled()) {
+        final StringBuilder buffer = new StringBuilder()
+                .append("Still waiting for: ");
+        int wrote = 0;
+        for (final Entry<QueryResultId, Waiter> entry : results.entrySet()) {
+          if (entry.getValue().result == null) {
+            if (wrote++ > 0) {
+              buffer.append(", ");
+            }
+            buffer.append(entry.getKey());
+          }
+        }
+        buffer.append(" on node ")
+              .append(config.getId());
+        LOG.trace(buffer.toString());
+      }
+
+      // start timers
+      for (final Entry<QueryResultId, Waiter> entry : results.entrySet()) {
+        if (entry.getKey().equals(next.dataSource())) {
+          continue;
+        }
+
+        final Waiter waiter = entry.getValue();
+        if (waiter.timeoutMillis > 0 && waiter.timeout == null) {
+          waiter.timeout = context.tsdb().getMaintenanceTimer().newTimeout(
+                  waiter, waiter.timeoutMillis, TimeUnit.MILLISECONDS);
+        }
+      }
+      return;
+    }
+
+    if (!complete.compareAndSet(false, true)) {
```

Comment:
Given that we already check complete.get() , is this needed ? Also, I assume the onNext method is called only by one thread at a time.

Reply:
My point is, if someone else other than this method is setting the complete variable to false on timeout.

Reply:
Yeah cause we have the race between the timeouts and data sources. The first `complete.get()` is just a double check to avoid doing any work building logs or starting timers in cases where the source finally comes in. The CAS is the final atomic gate. There's _still_ a potential to start extra timers but that should be pretty rare and would be caught by the CAS in the timeout.